### PR TITLE
1909-User model tests

### PR DESCRIPTION
**Pull request by MemoryBerg**

### CODE REVIEW

File: test/models/UserModel/test.js [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+      });
+      expect(error).to.be.exist;
+    });
+  });
+  describe('On find', async () => {
+    let name;
+    let user;
+    let userAlias;
+    let condition;
+    let usersCount;
```
Comment (olegtsigulov) [inline test/models/UserModel/test.js:231]: in one line

File: test/models/UserModel/test.js [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+      });
+      expect(Array.isArray(usersData)).to.be.true;
+    });
+    it('Should return correct user', async () => {
+      const { usersData } = await UserModel.find({
+        condition,
+        select: { alias: user.alias },
+        skip: 0,
+      });
+      expect(usersData[0].alias).to.be.eq(userAlias);
```
Comment (olegtsigulov) [inline test/models/UserModel/test.js:257]: you can make a nested describe method, in which, in a before hook, call the search method and in it cases, just do an expect

File: test/models/UserModel/test.js [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+    it('Should return an empty array if user not found', async () => {
+      const wrongName = faker.name.firstName();
+      const { usersData } = await UserModel.find({
+        condition: { name: wrongName },
+        select: { name: wrongName },
+        sort: 'name',
+        skip: 0,
+        limit: 10,
+      });
+      expect(_.isEmpty(usersData)).to.be.true;
```
Comment (olegtsigulov) [inline test/models/UserModel/test.js:268]: expect(usersData).to.be.empty;

File: test/models/UserModel/test.js [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+      await UsersFactory.Create();
+      const { result } = await UserModel.findWithSelect({ name: { $in: [new RegExp('^asdf_')] } },
+        'name, alias');
+      expect(result.length).to.be.eq(usersCount);
+    });
+    it('Should check that the error exists', async () => {
+      const { error } = await UserModel.findWithSelect([]);
+      expect(error).to.be.exist;
+    });
+  });
```
Comment (olegtsigulov) [inline test/models/UserModel/test.js:303]: you do not check that after the search, only those fields that you need are selected, and this is what this method was created for

File: test/models/UserModel/test.js [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    });
+    it('should return correct number', async () => {
+      const { count } = await UserModel.getCustomCount({});
+      expect(count).to.be.eq(usersCount);
+    });
+    it('Should return an error if condition transmitted incorrectly', async () => {
+      const { error } = await UserModel.getCustomCount(usersCount);
+      expect(error).to.be.exist;
+    });
+  });
```
Comment (olegtsigulov) [inline test/models/UserModel/test.js:321]: create conditions when not all users will be included in the sample, and check that the method correctly counted them

File: test/models/UserModel/test.js [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+describe('UserModel', () => {
+  describe('On getOne', () => {
+    let name;
+    beforeEach(async () => {
+      await dropDatabase();
+      name = faker.name.firstName();
+      await UsersFactory.Create({ name });
+    });
+    it('Should return the correct object', async () => {
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:15]: here yo check only name not object, rename or change expect

File: test/models/UserModel/test.js [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+      const { user } = await UserModel.getOne(name, ['name', 'wobjects_weight']);
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:16]: why you pass the second parameter to the method?

File: test/models/UserModel/test.js [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+      expect(user.name).to.be.eq(name);
+    });
+    it('Should check that the error exists', async () => {
+      const { error } = await UserModel.getOne({ host: faker.random.string() });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:20]: you just send empty obj to get error

File: test/models/UserModel/test.js [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+      expect(error).to.be.exist;
+    });
+    it('Should return null when user can\'t be found', async () => {
+      const { user: myUser } = await UserModel.getOne(faker.name.firstName());
+      expect(myUser).to.be.null;
+    });
+  });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:27]: add it to select(second param keys)

File: test/models/UserModel/test.js [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+      link = faker.random.string();
+      await UsersFactory.Create({ name, objects_follow: [link] });
+      await ObjectFactory.Create({ authorPermlink: link });
+      await UsersFactory.Create({ name: userName, objects_follow: [link] });
+    });
+    it('Should return an array', async () => {
+      const { wobjects } = await UserModel.getObjectsFollow({
+        name,
+      });
+      expect(Array.isArray(wobjects)).to.be.true;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:82]: expect([1, 2]).to.be.an('array')

File: test/models/UserModel/test.js [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    it('Should return empty array if username was not transmitted', async () => {
+      const { wobjects } = await UserModel.getObjectsFollow({ limit, skip: 0 });
+      expect(wobjects).to.be.empty;
+    });
+    it('Should check that array isn\'t empty', async () => {
+      const { wobjects } = await UserModel.getObjectsFollow({
+        name: userName,
+      });
+      expect(wobjects.length > 0).to.be.true;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:92]: expect([1]).to.be.an('array').that.is.not.empty;

File: test/models/UserModel/test.js [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+    it('Should return array with correct users.length', async () => {
+      await UsersFactory.Create();
+      const { result } = await UserModel.aggregate([
+        {
+          $match: {
+            name: new RegExp('^asdf_'),
+          },
+        }]);
+      expect(Array.isArray(result)).to.be.true;
+      expect(result.length).to.be.eq(usersCount);
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:123]: don't use two expect in one it, change Array.isArray, second expect change(result).to.have.length()

File: models/UserModel.js [hunk 13]
```diff
@@ -57,7 +58,7 @@ exports.aggregate = async (pipeline) => {
   try {
     const result = await UserModel.aggregate(pipeline).exec();
 
-    if (!result) {
+    if (_.isEmpty(result)) {
```
Comment (Fesmofet) [inline models/UserModel.js:61]: leave it(!result)

File: test/models/UserModel/test.js [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    it('Should return an error if pipeline is not correct', async () => {
+      const { error } = await UserModel.aggregate({ name: faker.name.firstName() });
+      expect(error).to.be.exist;
+      expect(error).to.be.an('error');
+    });
+    it('Should return an error', async () => {
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:130]: remove

File: test/models/UserModel/test.js [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+      );
+      expect(user.alias).to.be.eq(updateData);
+    });
+    it('Should change a field of user', async () => {
+      const { user } = await UserModel.updateOne(
+        { name },
+        { alias: updateData },
+      );
+      expect(user.alias).to.be.eq(updateData);
+    });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:157]: duplicate code, do you need first it?

File: test/models/UserModel/test.js [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+    it('Should change the privateEmail field of user', async () => {
+      const { user } = await UserModel.updateOne(
+        { privateEmail: true },
+        { privateEmail: false },
+      );
+      expect(user.privateEmail === 'false').to.be.true;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:163]: ???

File: test/models/UserModel/test.js [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+        notGuest: false,
+      };
+    });
+    it('Should return a correct user', async () => {
+      const { users } = await UserModel.search(options);
+      expect(users[0].name).to.be.eq(user.name);
+    });
+    it('Should return an array', async () => {
+      const { users } = await UserModel.search(options);
+      expect(Array.isArray(users)).to.be.true;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:191]: change

File: test/models/UserModel/test.js [hunk 18]
```diff
@@ -0,0 +1,328 @@
...
+    it('Should return an empty array if users was not found', async () => {
+      options.string = faker.random.word();
+      const { users } = await UserModel.search(options);
+      expect(users).to.be.empty;
+    });
+    it('Should check that the error exists', async () => {
+      const { error } = await UserModel.search({
+        test: _.random(0, 50),
+        host: faker.random.string(),
+        skip: user,
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:207]: skip string will be enough  for error

File: test/models/UserModel/test.js [hunk 19]
```diff
@@ -0,0 +1,328 @@
...
+      });
+      expect(error).to.be.exist;
+    });
+  });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:211]: create guest user and test search

File: test/models/UserModel/test.js [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+      let usersData;
+      beforeEach(async () => {
+        usersData = await UserModel.find({
+          condition,
+          select: { alias: userAlias },
+          skip: 0,
+        });
+      });
+      it('Should return an array', async () => {
+        expect(Array.isArray(usersData.usersData)).to.be.true;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:234]: change

File: test/models/UserModel/test.js [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    it('Should check that the error exists', async () => {
+      const { error } = await UserModel.find({
+        test: _.random(0, 50),
+        host: faker.random.string(),
+        skip: name,
+      });
+      expect(error).to.be.exist;
+    });
+  });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:259]: add tests on select and sort

File: test/models/UserModel/test.js [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+      expect(result.length).to.be.eq(usersCount);
+    });
+    it('Should return users with correct fields', async () => {
+      const randomUser = _.random(0, usersCount - 1);
+      await UsersFactory.Create();
+      const { result } = await UserModel.findWithSelect({ name: { $in: [new RegExp('^asdf_')] } },
+        'alias');
+      expect(result[randomUser].alias).to.be.eq('');
+      expect(result[randomUser].referral).to.be.undefined;
+      expect(result[randomUser].stage_version).to.be.undefined;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:287]: 1 expect for one it

File: test/models/UserModel/test.js [hunk 23]
```diff
@@ -0,0 +1,328 @@
...
+    });
+    it('Should return users with correct fields', async () => {
+      const alias = faker.random.string();
+      UsersFactory.Create({ name: `asdf_${faker.name.firstName()}`, alias });
+      const { result } = await UserModel.findWithSelect({ name: { $in: [new RegExp('^asdf_')] } },
+        { name: 1, alias: 1 });
+      expect(result[0].objects_follow).to.be.undefined;
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:294]: why you check this?

File: test/models/UserModel/test.js [hunk 24]
```diff
@@ -0,0 +1,328 @@
...
+      expect(count).to.be.eq(usersCount);
+    });
+    it('Should return a correct number of users', async () => {
+      const usersAliasCount = _.random(1, 10);
+      for (let iter = 0; iter < usersAliasCount; iter++) {
+        await UsersFactory.Create({ alias });
+      }
+      const { count } = await UserModel.getCustomCount({ alias });
+      expect(count).to.be.eq(usersAliasCount);
+    });
```
Comment (Fesmofet) [inline test/models/UserModel/test.js:322]: what's the difference between these two tests?

**Review by Fesmofet — COMMENTED** [review]
fix

**Review by Waiviogit — COMMENTED** [review]
fix

**Review by Fesmofet — APPROVED** [review]

**Pull request merged**